File: Homework/Ngan_Luu/202-happy-number.py
#https://leetcode.com/problems/happy-number/
class Solution:
    def isHappy(self, n: int) -> bool:
        number_dictionary = {}
        while(n != 1):
            digits = [int(i) for i in str(n)]
            n = 0
            for d in digits:
                n += d*d
            if n not in number_dictionary:
                number_dictionary[n] = 1
            else:
                return False
        return True

File: Homework/Son/W1/202-happy-number.js
/**
 * @param {number} n
 * @return {boolean}
 */
 var isHappy = function(n) {
    function getNext(num) {
        let res = 0;
        while (num > 0) {
            d = num % 10;
            num = parseInt(num/10);
            res += d * d;
        }
        return res;
    }
    let cycleSet = new Set()
    while (n !== 1) {
        if (!cycleSet.has(n)) cycleSet.add(n)
        else if (cycleSet.has(n)) return false
        n = getNext(n)
    }
    return true;
};

// https://leetcode.com/problems/happy-number/